### DISCUSSION TITLE
Clarify `--node-ip` vs `--node-public-ip` vs `--flannel-iface`

### DISCUSSION THREAD

**Discussion by intelfx** [k3s-io/k3s#9888] [Q&A]
Hi!

I'm (still) developing some automation around k3s management and it's not clear to me which network configurations a) are supported and b) translate to which combination of flags/configuration.

1. What exactly are `--node-ip` and `--node-external-ip` and when should I use either?
2. What is `--flannel-iface` and how does it relate to the two IP overrides above?

---

Or, more specifically: if I have two network interfaces on each node, where the first one (`eth1`) has a default route and the second one (`eth2`) is an isolated LAN connecting the nodes, is it possible to tell k3s to use the second interface for inter-node traffic and the first one only for cluster ingress/egress (because it might be less performant / more costly)?

Will the configuration differ depending on whether `eth1` directly has a globally routable IPv4 vs. a private IPv4 behind a NAT with forwarded ports?

---

Or, by example:

#### Case 1

* node1: eth1 = `1.2.3.4`, eth2 = `10.10.10.1/24`
* node2: eth1 = `5.6.7.8`, eth2 = `10.10.10.2/24`

(Assume that node1 may reach node2 by either IP, but the second one is preferable.)

#### Case 2

* node1: eth1 = `10.1.2.3` (NATed from `1.2.3.4`), eth2 = `10.10.10.1/24`
* node2: eth1 = `10.5.6.7` (NATed from `5.6.7.8`), eth2 = `10.10.10.2/24`

(Assume that node1 may reach node2 either by 5.6.7.8 or 10.10.10.2 (second one being preferable), but not by 10.5.6.7.)

Apologies for a long-winded question.

**Comment by brandond** [comment]
--node-ip sets the node's private IP, overriding the default behavior of selecting the primary IP of the interface targeted by the default route.
--node-external-ip sets the node's external IP, for cases where the node is behind 1:1 NAT and has an external address. In cloud provider environments, this is usually discovered by the cloud controller via instance metadata, but this allows it to be set by hand in unmanaged environments. It is unset by default.
--flannel-iface sets the interface used by flannel for network overlay traffic.

There is also the --flannel-external-ip flag, which when set on the server, configures flannel's behavior on all nodes in the cluster. Flannel will target the node's external IPs for network overlay traffic, instead of the private IP. This is intended for use in environments where the nodes cannot reach each other on their private IPs.

**Reply by intelfx** [reply]
> --node-ip sets the node's private IP, overriding the default behavior of selecting the primary IP of the interface targeted by the default route

In general, what should a node's private IP (and thus `--node-ip`) be set to (i.e. what's it used for)?

> node-external-ip sets the node's external IP, for cases where the node is behind 1:1 NAT and has an external address

Is it _required_ that `--node-external-ip` be set to the IP that is NATed to `--node-ip` (if any)? What will happen otherwise?

> --flannel-iface sets the interface used by flannel for network overlay traffic.

And if it's not set, then flannel will use the interface of `--node-ip`? What happens if it _is_ set — will flannel establish connections from `--flannel-iface` to other nodes' `--node-ip`, or to the respective IP discovered on `--flannel-iface` of other nodes?

---

Apologies once again, this all raises even more questions 😅

**Reply by brandond** [reply]
>  what should a node's private IP be set to

You don't have to set it at all, if you're happy with the default behavior of selecting the primary IP on the interface with the default route.

> Is it required that --node-external-ip be set

You only need to set it if you want K3s to be aware of the external IP, and be able to use it for things like loadbalancer or flannel destination addresses.

> And if it's not set, then flannel will use the interface of --node-ip

It will use the interface with that IP, correct.

> will flannel establish connections from --flannel-iface to other nodes

Yes, it sets the interface that flannel both initiates traffic from, and expects traffic to. These are really more flannel questions though...

**Reply by intelfx** [reply]
> You don't have to set it at all, if you're happy with the default behavior of selecting the primary IP on the interface with the default route.

I'm just trying to understand what it's _intended_ for -- i.e. what are the consequences of choosing one or another.

> if you're happy with the default behavior of selecting the primary IP on the interface with the default route.

My end goal is to configure k3s in such a way that all inter-node traffic goes via a specific isolated (i.e. non-default-route) interface, whereas the interface with the default route is only used for cluster ingress/egress.

I originally assumed that I could set `--node-ip` to the isolated LAN address and `--external-node-ip` to the public address, but if the latter is designed for 1:1 NAT cases, then I guess I need something else.

**Reply by appany-it** [reply]
@intelfx Did you find a solution to isolate overlay traffic?

**Reply by cytrinox** [reply]
@brandond 
> > And if it's not set, then flannel will use the interface of --node-ip
> 
> It will use the interface with that IP, correct.

I've tried this on my Hetzner nodes but flannel always picks the wrong interface. **eth0** is my public interface, **enp7s0** my private interface.

````
# ip -4 a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    inet 188.34.xxx.xxx/32 metric 100 scope global dynamic eth0
       valid_lft 48171sec preferred_lft 48171sec
3: enp7s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc fq_codel state UP group default qlen 1000
    inet 10.1.0.8/32 brd 10.1.0.8 scope global dynamic enp7s0
       valid_lft 48175sec preferred_lft 37375sec
````

My systemd k3s server command line is:
````
ExecStart=/usr/local/bin/k3s server --cluster-init --data-dir /var/lib/rancher/k3s --node-ip 10.1.0.8  --advertise-address 10.1.0.8 --tls-san 188.34.xxx.xxx --tls-san 2a01:4f8:1c1a:xxxx:xxxx --disable=traefik --tls-san 'admin-node.k3s.cluster.company.xyz' --cluster-domain stage.cluster.local --disable-cloud-controller --kubelet-arg cloud-provider=external
````

According to your answer, flannel should use enp7s0 because of the private node-ip 10.1.0.8. But it always picks eth0:

````
# ip -d link show type vxlan
20: flannel.1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc noqueue state UNKNOWN mode DEFAULT group default 
    link/ether 86:a1:41:3a:11:8a brd ff:ff:ff:ff:ff:ff promiscuity 0  allmulti 0 minmtu 68 maxmtu 65535 
    vxlan id 1 local 188.34.xxx.xxx dev eth0 srcport 0 0 dstport 8472 nolearning ttl auto ageing 300 udpcsum noudp6zerocsumtx noudp6zerocsumrx addrgenmode eui64 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 

````

I've debugged this behavior for days, because it's very subtle. You don't see in any logs that flannel is not working correctly. I hit this while deploying Hetzner CSI driver in my fresh cluster. All I got was:

    "Failed to connect to the CSI driver" err="context deadline exceeded" csiAddress="/run/csi/socket"

Turns out this was actually a DNS problem, because the CSI pod containers are unable to reach the Service CIDR 10.43.0.0/16 and core-dns has it service on 10.43.0.10. Without DNS, Hetzner CSI can't talk to the Hetzner Cloud API.

Has something changed in k3s/flannel or have I hit a bug?

Edit: k3s v1.32.3+k3s1 here, but it also happens with v1.30.x

**Reply by toabi** [reply]
I discovered this issue after hours finding out why after a k3s and os upgrade (k3s 1.31 to 1.32, and ubuntu just minor bump in kernel) our setup degraded. Luckily I noticed it early enough so currently it's in a "workaround"-state on "old" nodes. I'm also running k3s on Hetzner (configured with terraform).

New nodes added to the cluster apparently suddenly have a broken flannel-wireguard setup. It's not possible to access the overlay network from them. The rest is reachable as usual. It looks like it doesn't like the public IP addresses as wireguard endpoints anymore.

Setting the `--flannel-iface` might be a fix for that… which I am now exploring.
